packager/media/formats/mp2t/es_parser_teletext.cc

```diff
@@ -0,0 +1,518 @@
+std::string EsParserTeletext::BuildText(const uint8_t* data_block,
+                                        const uint8_t row) const {
+  const size_t payload_len = 40;
+  std::string next_string;
+  next_string.reserve(payload_len * 2);
+  bool leading_spaces = true;
+
+  const uint16_t index = magazine_ * 100 + page_number_;
+  const auto page_state_itr = page_state_.find(index);
+
+  const std::unordered_map<uint8_t, std::string>* column_replacement_map =
+      nullptr;
+  if (page_state_itr != page_state_.cend()) {
+    const auto row_itr =
+        page_state_itr->second.packet_26_replacements.find(row);
+    if (row_itr != page_state_itr->second.packet_26_replacements.cend()) {
+      column_replacement_map = &(row_itr->second);
+    }
+  }
+
+  for (size_t i = 0; i < payload_len; ++i) {
+    if (column_replacement_map) {
+      const auto column_itr = column_replacement_map->find(i);
+      if (column_itr != column_replacement_map->cend()) {
+        next_string.append(column_itr->second);
+        leading_spaces = false;
+        continue;
+      }
+    }
+
+    char next_char =
+        static_cast<char>(TELETEXT_BITREVERSE_8[data_block[i]] & 0x7f);
+
+    if (next_char < 32) {
+      next_char = 0x20;
+    }
+
+    if (leading_spaces) {
+      if (next_char == 0x20) {
+        continue;
+      }
+      leading_spaces = false;
+    }
+
+    switch (next_char) {
+      case '&':
+        next_string.append("&amp;");
+        break;
+      case '<':
+        next_string.append("&lt;");
+        break;
+      default: {
+        const std::string replacement(current_charset_[next_char - 0x20]);
+        next_string.append(replacement);
+      } break;
+    }
+  }
+
+  return RemoveTrailingSpaces(next_string);
+}
+
+void EsParserTeletext::ParsePacket26(const uint8_t* data_block) {
+  const uint16_t index = magazine_ * 100 + page_number_;
+  auto page_state_itr = page_state_.find(index);
+  if (page_state_itr == page_state_.end()) {
+    page_state_.emplace(index, TextBlock{{}, {}, last_pts_});
+  }
+  auto& replacement_map = page_state_[index].packet_26_replacements;
+
+  uint8_t row = 0;
+  const size_t payload_len = 40;
+
+  std::vector<uint32_t> x26_triplets;
+  x26_triplets.reserve(13);
+  for (uint8_t i = 1; i < payload_len; i += 3) {
+    const uint32_t bytes = (TELETEXT_BITREVERSE_8[data_block[i + 2]] << 16) |
+                           (TELETEXT_BITREVERSE_8[data_block[i + 1]] << 8) |
+                           TELETEXT_BITREVERSE_8[data_block[i]];
+    uint32_t triplet;
+    if (Hamming_24_18(bytes, triplet)) {
+      x26_triplets.emplace_back(triplet);
+    }
+  }
+
+  for (const auto triplet : x26_triplets) {
+    const uint8_t mode = (triplet & 0x7c0) >> 6;
+    const uint8_t address = triplet & 0x3f;
+    const uint8_t row_address_group = (address >= 0x28) && (address <= 0x3f);
+
+    if ((mode == 0x4) && (row_address_group == 0x1)) {
+      row = address - 0x28;
+      if (row == 0x0) {
+        row = 0x18;
+      }
+    }
```

Comment:
Named constants, please

Reply:
on the one hand I agree these need to be constants, on the other hand that might require someone familiar with the spec to come up with some meaningful names for them.